crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+use pyo3::types::{PyBool, PyDict, PySequence, PyString, PyTuple};
+use qiskit_circuit::bit_data::BitData;
+use qiskit_circuit::circuit_instruction::CircuitInstruction;
+use qiskit_circuit::dag_node::DAGOpNode;
+use qiskit_circuit::imports::QI_OPERATOR;
+use qiskit_circuit::operations::OperationRef::{Gate as PyGateType, Operation as PyOperationType};
+use qiskit_circuit::operations::{Operation, OperationRef, Param};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::distancemap::DistanceMap;
+
+const SKIPPED_NAMES: [&str; 4] = ["measure", "reset", "delay", "initialize"];
+const NO_CACHE_NAMES: [&str; 2] = ["annotated", "linear_function"];
+const SUPPORTED_OP: [&str; 18] = [
+    "h", "x", "y", "z", "sx", "sxdg", "t", "tdg", "s", "sdg", "cx", "cy", "cz", "swap", "iswap",
+    "ecr", "ccx", "cswap",
+];
+
+#[pyclass(module = "qiskit._accelerate.commutation_checker")]
+struct CommutationChecker {
+    library: CommutationLibrary,
+    cache_max_entries: usize,
+    cache: HashMap<(String, String), CommutationCacheEntry>,
+    #[pyo3(get)]
+    current_cache_entries: usize,
+    #[pyo3(get)]
+    _cache_miss: usize,
+    #[pyo3(get)]
+    _cache_hit: usize,
+    #[pyo3(get)]
+    gates: Option<HashSet<String>>,
+}
+
+#[pymethods]
+impl CommutationChecker {
+    #[pyo3(signature = (standard_gate_commutations=None, cache_max_entries=1_000_000, gates=None))]
+    #[new]
+    fn py_new(
+        py: Python,
+        standard_gate_commutations: Option<Py<PyAny>>,
+        cache_max_entries: usize,
+        gates: Option<HashSet<String>>,
+    ) -> Self {
+        CommutationChecker {
+            library: CommutationLibrary::new(py, standard_gate_commutations),
+            cache: HashMap::with_capacity(cache_max_entries),
+            cache_max_entries,
+            current_cache_entries: 0,
+            _cache_miss: 0,
+            _cache_hit: 0,
+            gates,
+        }
+    }
+
+    #[pyo3(signature=(op1, op2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute_nodes(
+        &mut self,
+        py: Python,
+        op1: &DAGOpNode,
+        op2: &DAGOpNode,
+        max_num_qubits: u32,
+    ) -> PyResult<bool> {
+        let mut bq: BitData<Qubit> = BitData::new(py, "qubits".to_string());
+        op1.instruction
+            .qubits
+            .bind(py)
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        op2.instruction
+            .qubits
+            .bind(py)
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        let qargs1 = op1
+            .instruction
+            .qubits
+            .bind(py)
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0 as usize)
+            .collect::<Vec<_>>();
+        let qargs2 = op2
+            .instruction
+            .qubits
+            .bind(py)
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0 as usize)
+            .collect::<Vec<_>>();
```

Comment:
There is a lot of unnecessary calls to `.bind(py)` here, you could just call this method once and store it in a variable.

```suggestion
        let mut bq: BitData<Qubit> = BitData::new(py, "qubits".to_string());
        let op1_bound_qubits = op1.instruction.qubits.bind(py);
        let op2_bound_qubits = op2.instruction.qubits.bind(py);
        op1_bound_qubits
            .iter()
            .for_each(|q| bq.add(py, &q, false).unwrap());
        op2_bound_qubits
            .iter()
            .for_each(|q| bq.add(py, &q, false).unwrap());
        let qargs1 = op1_bound_qubits
            .iter()
            .map(|q| bq.find(&q).unwrap().0 as usize)
            .collect::<Vec<_>>();
        let qargs2 = op2_bound_qubits
            .iter()
            .map(|q| bq.find(&q).unwrap().0 as usize)
            .collect::<Vec<_>>();
```